protocol/binary.go

```diff
@@ -61,3 +89,116 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	e, err := reader.ReadEnveloped()
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
+// specific purpose of decoding request structs that may or may not have an
+// envelope.
+// This allows a Thrift request handler to transparently accept requests
+// regardless of whether the caller submits an envelope.
+// The caller specifies the expected envelope type, one of OneWay or Unary, on
+// which the decoder asserts if the envelope is present.
+func (b binaryProtocol) DecodeRequest(et wire.EnvelopeType, r io.ReaderAt) (wire.Value, EnvelopeSpecificResponder, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
+	// without a version (non-strict)
+	// 3. A message of length >1 with MSB 0x80 set can only be enveloped with a
+	// version (strict)
+	// 4. A message of length >1 starting with 0x00-0x0f can only be the
+	// beginning of an un-enveloped field.
+	var buf [2]byte
+	// If we fail to read two bytes, the only possible valid value is the empty struct.
+	if _, err := r.ReadAt(buf[0:2], 0); err != nil {
```

Comment:
Instead of relying on `err != nil`, it may be preferable to rely on the first return value of `ReadAt`. (We're ignoring the error anyway.)

```lang=go
if n, _ := r.ReadAt(buf[:2], 0); n < 2 {
    ...
}
```